File: tundra-cli/benchmarks/sieve/sieve.java

public class sieve {
    public static int countPrimes(int n) {
        boolean[] isComposite = new boolean[n + 1];
        int p = 2;
        while (p * p <= n) {
            if (!isComposite[p]) {
                for (int m = p * p; m <= n; m += p) {
                    isComposite[m] = true;
                }
            }
            p++;
        }
        int count = 0;
        for (int i = 2; i <= n; i++) {
            if (!isComposite[i]) count++;
        }
        return count;
    }

    public static void main(String[] args) {
        int n = 200000;
        if (args.length > 0) {
            n = Integer.parseInt(args[0]);
        }
        System.out.println(countPrimes(n));
    }
}

File: tundra-cli/benchmarks/sieve/sieve.nim
# File: sieve/sieve.nim
import sequtils, strutils, os

proc sieve(n: int): int =
  var isComposite = newSeq[bool](n + 1)
  for i in 0..n:
    isComposite[i] = false

  var p = 2
  while p * p <= n:
    if not isComposite[p]:
      var m = p * p
      while m <= n:
        isComposite[m] = true
        m += p
    p.inc()
  
  # Count primes:
  result = 0
  for i in 2..n:
    if not isComposite[i]:
      result.inc()

when isMainModule:
  var n = 200000
  if paramCount() > 0:
    n = parseInt(paramStr(1))
  echo sieve(n)
